Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                assert(self.waiters.isEmpty, "If a connection is available, why are there any waiters")
+                var connectionState = self.connections[index]
+                let connection = connectionState.lease()
+                self.connections[index] = connectionState
+                return .init(
+                    .executeRequest(request, connection, cancelWaiter: nil),
+                    .cancelTimeoutTimer(connectionState.connectionID)
+                )
+            }
+
+            // if we found a good match, let's use this
+            if let (index, _) = goodMatch {
+                assert(self.waiters.isEmpty, "If a connection is available, why are there any waiters")
+                if !required {
+                    var connectionState = self.connections[index]
+                    let connectionID = connectionState.connectionID
+                    let connection = connectionState.lease()
+                    self.connections[index] = connectionState
+                    return .init(
+                        .executeRequest(request, connection, cancelWaiter: nil),
+                        .cancelTimeoutTimer(connectionID)
+                    )
+                } else {
+                    assert(self.connections.count - self.maximumConcurrentConnections == 0)
+                    var oldConnectionState = self.connections[index]
+                    let newConnectionID = self.idGenerator.next()
+                    let newWaiter = Waiter(request: request)
+                    self.connections[index] = .init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter)
+                    return .init(
+                        .scheduleWaiterTimeout(newWaiter.requestID, request, on: prefferedEL),
+                        .replaceConnection(oldConnectionState.close(), with: newConnectionID, on: prefferedEL)
+                    )
+                }
+            }
+
+            // we didn't find any match at all... Let's create a new connection, if there is room
+            // left
+            if self.connections.count < self.maximumConcurrentConnections {
+                let newConnectionID = self.idGenerator.next()
+                let newWaiter = Waiter(request: request)
+                self.connections.append(.init(connectionID: newConnectionID, eventLoop: prefferedEL, waiter: newWaiter))
+                return .init(
+                    .scheduleWaiterTimeout(newWaiter.requestID, request, on: prefferedEL),
+                    .createConnection(newConnectionID, on: prefferedEL)
+                )
+            }
+
+            // all connections are busy, and there is no more room to create further connections
+            let waiter = Waiter(request: request)
+            self.waiters.append(waiter)
+            return .init(
+                .scheduleWaiterTimeout(waiter.requestID, request, on: prefferedEL),
+                .none
+            )
+        }
```

Comment:
This method is 140 lines long. It needs to be refactored into several smaller methods or it will be impossible to follow.

Reply:
+1